engine/resources/hetzner.go

```diff
@@ -0,0 +1,716 @@
+	Image string `lang:"image"`
+
+	// State is the desired steady state of the server instance.
+	// Choose between "running", "off" or "absent".
+	State string `lang:"state"`
+
+	// WaitInterval is the sleep interval in seconds that is used when waiting
+	// for transient states to converge between intermediate operations.
+	// Note that the hetzner API supports bursts at up to 3600 requests per hour.
+	// Nonzero values disable request bursts, a zero value will allow it.
+	// A wait interval >= 1s is recommended for extended use.
+	WaitInterval uint32 `lang:"waitinterval"`
+
+	// ServerRescueMode specifies the image used when enabling rescue mode.
+	// The supported images are "linux32", "linux64" and "freebsd64".
+	// Alternatively, use "disabled" to disable rescue mode (default).
+	ServerRescueMode string `lang:"serverrescuemode"`
+	// ServerRescueSSHKeys is used to select a subset of keys that
+	// should be available for rescue mode operations over ssh.
+	// From all ssh keys known to the project client, choose a subset of
+	// keys by name, as an array of strings. New keys need to be added
+	// manually via the cloud console web api for now.
+	// If a keyname is unknown to the client, it is ignored.
+	ServerRescueSSHKeys []string `lang:"serverrescuekeys"`
+
+	// Full client details for interaction with the hcloud-go API.
+	client *hcloud.Client
+	// Snapshot of the server instance, for checkapply operations.
+	server *hcloud.Server
+	// Snapshot of the serverCreateOpts struct generated with hcloud.
+	serverconfig hcloud.ServerCreateOpts
+	// hcloud-formatted array of ssh keys for use with rescue mode.
+	rescuekeys []*hcloud.SSHKey
+
+	// Internal flag to simplify rescue mode operations in checkapply.
+	// False if ServerRescueMode="disabled", true otherwise.
+	rescueEnable bool
+}
+
+func (obj *HetznerRes) Default() engine.Res {
+	return &HetznerRes{
+		WaitInterval:     WaitIntervalLimit,
+		ServerRescueMode: "disabled",
+	}
+}
+
+func (obj *HetznerRes) Validate() error {
+
+	// verify name (TODO: how to ensure unique names across server instances?)
+	if obj.Name() == "" {
+		return fmt.Errorf("instance has empty name field")
+	}
+
+	// verify token (source and contents)
+	content, err := readFileLines(obj.APITokenPath)
+	if err != nil {
+		return errwrap.Wrapf(err, "no valid file in APITokenPath")
+	}
+	if content == nil {
+		return fmt.Errorf("file in APITokenPath is empty")
+	}
+	if len(content) > 1 {
+		return fmt.Errorf("file in APITokenPath contains more than one line")
+	}
+
+	// check polling interval
+	if obj.MetaParams().Poll < HetznerPollLimit {
+		return fmt.Errorf("invalid polling interval (min = %d s)", HetznerPollLimit)
+	}
+
+	// check wait interval
+	if obj.WaitInterval < WaitIntervalLimit {
+		return fmt.Errorf("invalid wait interval (min = %d)", WaitIntervalLimit)
+	}
+
+	// verify state specifiers (only allow steady states)
+	switch obj.State {
+	case StateRunning, StateOff, StateAbsent:
+		// continue, state is valid.
```

Comment:
Comments are either full sentences and are capitalized and end with proper punctuation, or are short comments that don't end in a period. Please correct all of these throughout. Larger guide here: https://github.com/purpleidea/mgmt/blob/master/docs/style-guide.md